File: Atracciones.Frontend/atraxia/src/app/services/auth.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import api from './api.service';

export interface AuthUser { username: string; token: string; expiration: string; roles: string[]; }

@Injectable({ providedIn: 'root' })
export class AuthService {
  private userSubject = new BehaviorSubject<AuthUser | null>(
    JSON.parse(localStorage.getItem('atraxia_user') || 'null')
  );
  currentUser$ = this.userSubject.asObservable();
  get currentUser() { return this.userSubject.value; }
  get isLoggedIn() { return !!this.userSubject.value; }

  async login(payload: { login: string; password: string }) {
    try {
      const res = await api.post('/Auth/login', payload);
      const data = res.data?.data;
      if (data?.success && data?.token) {
        const user: AuthUser = { username: data.username, token: data.token, expiration: data.expiration, roles: data.roles || [] };
        localStorage.setItem('atraxia_token', data.token);
        localStorage.setItem('atraxia_user', JSON.stringify(user));
        this.userSubject.next(user);
        return { success: true, message: `Bienvenido, ${data.username}` };
      }
      return { success: false, message: data?.message || 'Credenciales incorrectas' };
    } catch (err: any) {
      return { success: false, message: err.response?.data?.message || 'Error al iniciar sesión' };
    }
  }

  async register(payload: any) {
    try {
      const res = await api.post('/Auth', payload);
      return res.data?.success || res.data?.Success
        ? { success: true, message: 'Cuenta creada. Ya puedes ingresar.' }
        : { success: false, message: res.data?.message || res.data?.Message || 'Error al registrarse', errors: res.data?.errors || res.data?.Errors };
    } catch (err: any) {
      const data = err.response?.data;
      return { success: false, message: data?.message || data?.Message || 'Error al registrarse', errors: data?.errors || data?.Errors };
    }
  }

  async adminLogin(payload: { login: string, password: string }): Promise<{ success: boolean; message: string }> {
    try {
      const res = await api.post('/Auth/login-admin', payload);
      const data = res.data?.data;
      if (!data?.success || !data?.token) {
        return { success: false, message: data?.message || 'Credenciales incorrectas' };
      }
      localStorage.setItem('atraxia_admin_token', data.token);
      localStorage.setItem('atraxia_admin_user', data.username);

      return { success: true, message: 'Bienvenido al panel de administración' };
    } catch (err: any) {
      if (err.response?.status === 403) {
        return { success: false, message: 'Acceso denegado: no tienes permisos de administrador' };
      }
      return { success: false, message: 'No se pudo validar el acceso de administrador' };
    }
  }

  get isAdmin(): boolean {
    return !!localStorage.getItem('atraxia_admin_token');
  }

  get adminUsername(): string {
    return localStorage.getItem('atraxia_admin_user') || '';
  }

  adminLogout(): void {
    localStorage.removeItem('atraxia_admin_token');
    localStorage.removeItem('atraxia_admin_user');
  }

  logout() {
    localStorage.removeItem('atraxia_token');
    localStorage.removeItem('atraxia_user');
    this.userSubject.next(null);
  }
}

File: Atracciones.Frontend/atraxia_mobile/lib/services/auth_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../model/auth_user.dart';
import 'api_service.dart';

class AuthService {
  final ApiService _api = ApiService();

  Future<bool> get isLoggedIn async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('atraxia_token') != null;
  }

  Future<AuthUser?> get currentUser async {
    final prefs = await SharedPreferences.getInstance();
    final userStr = prefs.getString('atraxia_user');
    if (userStr != null) {
      return AuthUser.fromJson(jsonDecode(userStr));
    }
    return null;
  }

  Future<Map<String, dynamic>> login(Map<String, dynamic> payload) async {
    try {
      final res = await _api.post('/Auth/login', payload);
      final body = jsonDecode(res.body);
      final data = body['data'];
      
      if (res.statusCode >= 200 && res.statusCode < 300 && data != null && data['success'] == true) {
        final token = data['token'];
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString('atraxia_token', token);
        
        final user = AuthUser(
          username: data['username'] ?? '',
          token: token,
          expiration: data['expiration'] ?? '',
          roles: List<String>.from(data['roles'] ?? []),
        );
        await prefs.setString('atraxia_user', jsonEncode(user.toJson()));
        
        return {'success': true, 'message': 'Bienvenido, ${data['username']}'};
      }
      return {'success': false, 'message': data?['message'] ?? body['message'] ?? 'Credenciales incorrectas'};
    } catch (err) {
      return {'success': false, 'message': 'Error al iniciar sesión'};
    }
  }

  Future<Map<String, dynamic>> register(Map<String, dynamic> payload) async {
    try {
      final res = await _api.post('/Auth', payload);
      final body = jsonDecode(res.body);
      
      if (res.statusCode >= 200 && res.statusCode < 300 && (body['success'] == true || body['Success'] == true)) {
        return {'success': true, 'message': 'Cuenta creada. Ya puedes ingresar.'};
      }
      return {
        'success': false, 
        'message': body['message'] ?? body['Message'] ?? 'Error al registrarse',
        'errors': body['errors'] ?? body['Errors']
      };
    } catch (err) {
      return {'success': false, 'message': 'Error al registrarse'};
    }
  }

  Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('atraxia_token');
    await prefs.remove('atraxia_user');
  }
}
